# Day 6 Task 2: AFL Players

### Stuff to Cover
1. Data Quality Assessment
2. Cleaning
3. Merging and Unmatched ID stuff
4. Validation repport
   

## Imports and Load

In [3]:
import pandas as pd
import numpy as np

In [4]:
info_df = pd.read_csv("afl_players_info_raw.csv")
stats_df = pd.read_csv("afl_players_seasonal_stats_raw.csv")


C:\Users\imama\AppData\Local\Temp\ipykernel_3364\621726609.py:2: DtypeWarning: Columns (0: player_id) have mixed types. Specify dtype option on import or set low_memory=False.
  stats_df = pd.read_csv("afl_players_seasonal_stats_raw.csv")


In [5]:
info_df

,id,player_name,player_full_name,first_name,last_name,born_date,debut_date,debut_age,last_date,last_age,height,weight,profile_pic,player_link,player_common_names,player_teams
0,43261,Ryan Abbott,Ryan_Abbott,Ryan,Abbott,1991-06-25,2018-08-02,27,2020-09-05,29,200,100,NaN,https://afltables.com/afl/stats/players/R/Ryan...,NaN,"{Geelong Cats,St Kilda Saints}"
1,43262,Gary Ablett,Gary_Ablett1,Gary,Ablett,1984-05-14,2002-03-30,17,2020-10-24,36,182,87,NaN,https://afltables.com/afl/stats/players/G/Gary...,NaN,"{Geelong Cats,Gold Coast Suns}"
2,43276,Leek Aleer,Leek_Aleer,Leek,Aleer,2001-08-21,2022-07-30,20,2025-09-06,24,195,85,https://res.cloudinary.com/dijzdikkh/image/upl...,https://afltables.com/afl/stats/players/L/Leek...,NaN,NaN
3,43264,Nathan Ablett,Nathan_Ablett,Nathan,Ablett,1985-12-13,2005-08-12,19,2011-08-28,25,195,95,NaN,https://afltables.com/afl/stats/players/N/Nath...,NaN,"{Geelong Cats,Gold Coast Suns}"
4,43265,Cain Ackland,Cain_Ackland,Cain,Ackland,1982-03-16,2001-03-30,19,2008-05-16,26,196,101,NaN,https://afltables.com/afl/stats/players/C/Cain...,NaN,"{Carlton Blues,Port Adelaide Power,St Kilda Sa..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2843,43659,Charlie Curnow,Charlie_Curnow,charlie,CURNOW,1997-02-03,2016-04-03,19,2025-07-24,28,192,96,https://res.cloudinary.com/dijzdikkh/image/upl...,https://afltables.com/afl/stats/players/C/Char...,NaN,{Carlton Blues}
2844,44352,Matt Maguire,Matt_Maguire,matt,MAGUIRE,1984-05-30,2002-05-03,17,2015-04-18,30,192,101,NaN,https://afltables.com/afl/stats/players/M/Matt...,NaN,"{Brisbane Lions,St Kilda Saints}"
2845,45045,Cam Sutcliffe,Cam_Sutcliffe,cam,SUTCLIFFE,1992-05-23,2012-07-08,20,2020-09-05,28,186,87,NaN,https://afltables.com/afl/stats/players/C/Cam_...,NaN,"{Fremantle Dockers,Port Adelaide Power}"
2846,45738,Mark McVeigh,Mark_McVeigh,mark,MCVEIGH,1981-01-26,1999-03-25,18,2012-05-26,31,184,85,NaN,https://afltables.com/afl/stats/players/M/Mark...,NaN,{Essendon Bombers}


In [6]:
info_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2848 entries, 0 to 2847
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   id                   2848 non-null   int64
 1   player_name          2848 non-null   str  
 2   player_full_name     2848 non-null   str  
 3   first_name           2848 non-null   str  
 4   last_name            2848 non-null   str  
 5   born_date            2848 non-null   str  
 6   debut_date           2848 non-null   str  
 7   debut_age            2848 non-null   int64
 8   last_date            2848 non-null   str  
 9   last_age             2848 non-null   int64
 10  height               2848 non-null   int64
 11  weight               2848 non-null   int64
 12  profile_pic          637 non-null    str  
 13  player_link          2848 non-null   str  
 14  player_common_names  75 non-null     str  
 15  player_teams         2754 non-null   str  
dtypes: int64(5), str(11)
memory usage: 

In [7]:
info_df.isnull().sum()

id                        0
player_name               0
player_full_name          0
first_name                0
last_name                 0
born_date                 0
debut_date                0
debut_age                 0
last_date                 0
last_age                  0
height                    0
weight                    0
profile_pic            2211
player_link               0
player_common_names    2773
player_teams             94
dtype: int64

In [8]:
stats_df

,player_id,year,team,is_finals,games_played,kicks,marks,handballs,disposals,goals,...,avg_contested_possessions,avg_uncontested_possessions,avg_contested_marks,avg_marks_inside_50,avg_one_percenters,avg_bounces,avg_goal_assists,avg_score,avg_fantasy_points,avg_percentage_played
0,43261,2018,Geelong Cats,False,3,17.0,9.0,21.0,38.0,3.0,...,6.7,6.7,0.3,1.7,4.3,0.0,0.3,6.7,92.7,84.3
1,43261,2018,Geelong Cats,True,1,3.0,2.0,2.0,5.0,0.0,...,2.0,3.0,1.0,1.0,5.0,0.0,0.0,1.0,61.0,81.0
2,43261,2019,Geelong Cats,False,1,5.0,3.0,6.0,11.0,1.0,...,6.0,8.0,0.0,1.0,3.0,0.0,0.0,6.0,74.0,83.0
3,43261,2020,St Kilda Saints,False,1,2.0,2.0,1.0,3.0,1.0,...,0.0,3.0,0.0,1.0,0.0,0.0,0.0,6.0,22.0,54.0
4,43262,2002,Geelong Cats,False,12,37.0,13.0,63.0,100.0,10.0,...,5.7,3.0,0.3,0.3,1.0,0.2,0.0,5.3,36.9,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
25486,ID_43738,2013,Greater Western Sydney Giants,False,1,3.0,1.0,1.0,4.0,0.0,...,2.0,1.0,0.0,0.0,3.0,0.0,0.0,0.0,30.0,85.0
25487,ID_43660,2015,Carlton Blues,False,22,216.0,75.0,262.0,478.0,0.0,...,8.4,13.2,0.3,0.0,1.2,0.2,0.2,0.1,79.6,90.5
25488,ID_46153,2025,Melbourne Demons,False,5,15.0,5.0,12.0,27.0,3.0,...,3.4,1.8,0.4,0.8,2.8,0.0,0.4,4.2,30.8,62.4
25489,ID_44965,2005,Fremantle Dockers,False,6,30.0,11.0,41.0,71.0,1.0,...,4.8,7.2,0.0,0.2,1.8,0.2,0.0,1.3,41.5,47.8


In [9]:
stats_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 25491 entries, 0 to 25490
Data columns (total 54 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   player_id                    25491 non-null  object 
 1   year                         25491 non-null  int64  
 2   team                         25491 non-null  str    
 3   is_finals                    25491 non-null  bool   
 4   games_played                 25491 non-null  int64  
 5   kicks                        25439 non-null  float64
 6   marks                        25306 non-null  float64
 7   handballs                    25378 non-null  float64
 8   disposals                    25467 non-null  float64
 9   goals                        22954 non-null  float64
 10  behinds                      22802 non-null  float64
 11  hit_outs                     19785 non-null  float64
 12  tackles                      24980 non-null  float64
 13  rebound_50s                

In [10]:
stats_df.isnull().sum()

player_id                         0
year                              0
team                              0
is_finals                         0
games_played                      0
kicks                            52
marks                           185
handballs                       113
disposals                        24
goals                          2537
behinds                        2689
hit_outs                       5706
tackles                         511
rebound_50s                    3533
inside_50s                     2971
clearances                     3435
clangers                       2938
free_kicks_for                 1073
free_kicks_against              978
brownlow_votes                 6997
contested_possessions          3336
uncontested_possessions        3322
contested_marks                4925
marks_inside_50                5165
one_percenters                 3643
bounces                        5654
goal_assists                   7092
total_score                 

looks like we got a lotta cleaning to do 

## 1. Data Quality Assessment 
Gotta check for:
- dupes
- missing vals
- invalid vals (e.g. 0 weight)
- inconsistent logic
- messed up weight/height
#### info_df


In [11]:
#shape
info_df.shape

(2848, 16)

In [12]:
# dupe rows
info_df.duplicated().sum()

np.int64(5)

Found 5 dupes

In [14]:
#missing vals
info_df.isnull().sum()

id                        0
player_name               0
player_full_name          0
first_name                0
last_name                 0
born_date                 0
debut_date                0
debut_age                 0
last_date                 0
last_age                  0
height                    0
weight                    0
profile_pic            2211
player_link               0
player_common_names    2773
player_teams             94
dtype: int64

In [15]:
#impossible vals
print("rows with weight of 0")
info_df[info_df['weight'] == 0][['id', 'player_name', 'height', 'weight']]

rows with weight of 0


,id,player_name,height,weight
1824,45612,Jack Hutchinson,190,0
2228,46122,Tom Hanily,179,0


Found 2 rows that have weight of 0

In [16]:
#height and weight
info_df['height'].describe()

count    2848.000000
mean      187.344803
std         7.318328
min       163.000000
25%       182.000000
50%       187.000000
75%       192.000000
max       211.000000
Name: height, dtype: float64

In [17]:
info_df['weight'].describe()

count    2848.000000
mean       86.012289
std         8.923524
min         0.000000
25%        80.000000
50%        85.000000
75%        92.000000
max       118.000000
Name: weight, dtype: float64

## info_df Findings

# Data Quality Assessment Report of `info_df`

#### 1. Duplicated Rows: 5
I found 5 duplicate rows in the dataset where the same player ID and all information were repeated.

#### 2. Weight == 0: 2 found
There were 2 players with a weight of 0, which is not a valid value because a person's weight cannot be zero.

#### 3. Missing profile_pic: 2211
The profile_pic column had 2,211 missing values. Since this is only an optional profile picture field, the missing values were left as they are, we dont need to touch that.

#### 4. Missing player_common_names: 2773
The player_common_names column had 2,773 missing values. This field stores player nicknames, so leaving these values blank does not affect the analysis either

#### 5. Missing player_teams: 94
The player_teams column had 94 missing values. This means some players do not have a recorded team list, which is acceptable.

There were also no unrealistic height or weight values. Although some players were very short (163 cm) or very tall (211 cm), these are real players and not data errors.

#### stats_df

In [18]:
stats_df

,player_id,year,team,is_finals,games_played,kicks,marks,handballs,disposals,goals,...,avg_contested_possessions,avg_uncontested_possessions,avg_contested_marks,avg_marks_inside_50,avg_one_percenters,avg_bounces,avg_goal_assists,avg_score,avg_fantasy_points,avg_percentage_played
0,43261,2018,Geelong Cats,False,3,17.0,9.0,21.0,38.0,3.0,...,6.7,6.7,0.3,1.7,4.3,0.0,0.3,6.7,92.7,84.3
1,43261,2018,Geelong Cats,True,1,3.0,2.0,2.0,5.0,0.0,...,2.0,3.0,1.0,1.0,5.0,0.0,0.0,1.0,61.0,81.0
2,43261,2019,Geelong Cats,False,1,5.0,3.0,6.0,11.0,1.0,...,6.0,8.0,0.0,1.0,3.0,0.0,0.0,6.0,74.0,83.0
3,43261,2020,St Kilda Saints,False,1,2.0,2.0,1.0,3.0,1.0,...,0.0,3.0,0.0,1.0,0.0,0.0,0.0,6.0,22.0,54.0
4,43262,2002,Geelong Cats,False,12,37.0,13.0,63.0,100.0,10.0,...,5.7,3.0,0.3,0.3,1.0,0.2,0.0,5.3,36.9,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
25486,ID_43738,2013,Greater Western Sydney Giants,False,1,3.0,1.0,1.0,4.0,0.0,...,2.0,1.0,0.0,0.0,3.0,0.0,0.0,0.0,30.0,85.0
25487,ID_43660,2015,Carlton Blues,False,22,216.0,75.0,262.0,478.0,0.0,...,8.4,13.2,0.3,0.0,1.2,0.2,0.2,0.1,79.6,90.5
25488,ID_46153,2025,Melbourne Demons,False,5,15.0,5.0,12.0,27.0,3.0,...,3.4,1.8,0.4,0.8,2.8,0.0,0.4,4.2,30.8,62.4
25489,ID_44965,2005,Fremantle Dockers,False,6,30.0,11.0,41.0,71.0,1.0,...,4.8,7.2,0.0,0.2,1.8,0.2,0.0,1.3,41.5,47.8


Seems like there are rows here with- weird non numeric stuff


In [19]:
#player_id inconsistency
non_numeric = stats_df[
    ~stats_df['player_id'].astype(str).str.fullmatch(r'\d+', na=False)]


print("Number of non-numeric player_id values:", len(non_numeric))
print("\nNon-numeric player_id values:")

print(non_numeric['player_id'].unique())


Number of non-numeric player_id values: 10

Non-numeric player_id values:
['ID_44242' 'ID_44493' 'ID_45491' 'ID_44358' 'ID_44754' 'ID_43738'
 'ID_43660' 'ID_46153' 'ID_44965' 'ID_45608']


There seem to be 10 non numeric player ids

In [20]:
#duplicated rows
stats_df.duplicated().sum()

np.int64(10)

Found 10 duplicated rows

In [21]:
# missing values
stats_df.isnull().sum()

player_id                         0
year                              0
team                              0
is_finals                         0
games_played                      0
kicks                            52
marks                           185
handballs                       113
disposals                        24
goals                          2537
behinds                        2689
hit_outs                       5706
tackles                         511
rebound_50s                    3533
inside_50s                     2971
clearances                     3435
clangers                       2938
free_kicks_for                 1073
free_kicks_against              978
brownlow_votes                 6997
contested_possessions          3336
uncontested_possessions        3322
contested_marks                4925
marks_inside_50                5165
one_percenters                 3643
bounces                        5654
goal_assists                   7092
total_score                 

um lets sort that

In [22]:
stats_df.isnull().sum().sort_values(ascending = False)

avg_goal_assists               7092
goal_assists                   7092
brownlow_votes                 6997
avg_percentage_played          6246
total_percentage_played        6246
hit_outs                       5706
avg_hit_outs                   5706
avg_bounces                    5654
bounces                        5654
marks_inside_50                5165
avg_marks_inside_50            5165
contested_marks                4925
avg_contested_marks            4925
avg_one_percenters             3643
one_percenters                 3643
rebound_50s                    3533
avg_rebound_50s                3533
avg_clearances                 3435
clearances                     3435
avg_contested_possessions      3336
contested_possessions          3336
uncontested_possessions        3322
avg_uncontested_possessions    3322
avg_inside_50s                 2971
inside_50s                     2971
avg_clangers                   2938
clangers                       2938
behinds                     

Well thats a lot

In [23]:
# neg values where they shouldn't be possible
numcols = stats_df.select_dtypes('number').columns
neg_counts = (stats_df[numcols] < 0).sum()
neg_counts[neg_counts > 0]


games_played             8
total_fantasy_points    10
avg_fantasy_points      10
dtype: int64

Thats a lotta neg counts

In [24]:
print("Negative games_played rows:")
stats_df[stats_df['games_played'] < 0][['player_id','year','team','games_played','total_fantasy_points']]


Negative games_played rows:


,player_id,year,team,games_played,total_fantasy_points
19642,43260,2020,Richmond Tigers,-13,567
19643,43260,2020,Richmond Tigers,-1,45
25030,43260,2021,Richmond Tigers,-21,996
25031,43260,2022,Richmond Tigers,-7,199
25032,43260,2020,Richmond Tigers,-13,567
25033,43260,2020,Richmond Tigers,-1,45
25034,43260,2021,Richmond Tigers,-21,996
25035,43260,2022,Richmond Tigers,-7,199


In [25]:
# team nme inconsistency as in case + whitespace variants of the same team
print("Distinct raw team string values:", stats_df['team'].nunique())
sorted(stats_df['team'].dropna().unique())


Distinct raw team string values: 114


[' ADELAIDE CROWS ',
 ' Adelaide Crows ',
 ' BRISBANE LIONS ',
 ' Brisbane Lions ',
 ' CARLTON BLUES ',
 ' COLLINGWOOD MAGPIES ',
 ' Carlton Blues ',
 ' Collingwood Magpies ',
 ' ESSENDON BOMBERS ',
 ' Essendon Bombers ',
 ' FREMANTLE DOCKERS ',
 ' Fitzroy Lions ',
 ' Fremantle Dockers ',
 ' GOLD COAST SUNS ',
 ' GREATER WESTERN SYDNEY GIANTS ',
 ' Geelong Cats ',
 ' Gold Coast Suns ',
 ' Greater Western Sydney Giants ',
 ' HAWTHORN HAWKS ',
 ' Hawthorn Hawks ',
 ' Melbourne Demons ',
 ' NORTH MELBOURNE KANGAROOS ',
 ' North Melbourne Kangaroos ',
 ' PORT ADELAIDE POWER ',
 ' Port Adelaide Power ',
 ' RICHMOND TIGERS ',
 ' Richmond Tigers ',
 ' ST KILDA SAINTS ',
 ' SYDNEY SWANS ',
 ' St Kilda Saints ',
 ' Sydney Swans ',
 ' WEST COAST EAGLES ',
 ' WESTERN BULLDOGS ',
 ' West Coast Eagles ',
 ' Western Bulldogs ',
 ' adelaide crows ',
 ' brisbane bears ',
 ' brisbane lions ',
 ' carlton blues ',
 ' collingwood magpies ',
 ' essendon bombers ',
 ' fremantle dockers ',
 ' geelong cats ',

In [26]:
# Year range hceck

print("Year range:", stats_df['year'].min(), "to", stats_df['year'].max())


Year range: 1983 to 2025


## stats_df Findings


# Data Quality Assessment Report of `stats_df`

#### 1. player_id format inconsistency : 10 rows found
I found **10 `player_id` values** that were written as `ID_xxxxx` instead of just numbers. These were converted to a numeric format so the datasets could be merged correctly.

#### 2. Duplicated rows: 10
There were **10 duplicate rows** where the same player statistics were repeated. These duplicate records were removed.

#### 3. Negative rows for `game_played` : 4
I found **4 rows with negative `games_played` values**, which is not possible. These were identified as data entry errors and corrected during cleaning.

#### 4. Negative rows for `total_fantasy_points` : 10

There were **10 rows with negative `total_fantasy_points` values**. These were **not changed** because fantasy points can sometimes be negative based on the scoring system.

#### 5. `team` inconsistency: 114 raw str variants, had mixed case
The `team` column had inconsistent names due to different letter cases and extra spaces. These were standardized so each team has a consistent name.

#### 6. Missing val columns: 45
Missing values were found in **45 out of 54 columns**, mostly in advanced statistics such as `hit_outs`, `brownlow_votes`, and `contested_possessions`. These values were left unchanged because these statistics were not recorded for every player or season.

The data covers the years **1983 to 2025**, and all year values were valid, so no changes were needed.


# 2. Cleaning
### `info_df`
#### Decisions
1. **Drop exact duplicate rows**
2. **`weight == 0` → `NaN`** : zero kilograms is not a real value for an AFL player; treating it as missing avoids polluting any weight-based analysis while preserving the rest of the record.
4. **Parse date columns to `datetime`**:  enables correct date arithmetic and consistent downstream formatting.
5. **Leave `profile_pic`, `player_common_names`, `player_teams` missing values as `NaN`** : these are optional/metadata fields; imputing them would fabricate data (e.g. inventing a nickname or team history).
6. **No changes to height/weight extremes or date logic** : verified as genuine values, not errors.


In [27]:
info = info_df.copy()

# 1. Remove exact duplicate rows
dup_mask = info.duplicated()
n_dup = dup_mask.sum()
info = info[~dup_mask].copy()
print(f"Removed {n_dup} exact duplicate rows.")

# 2. weight == 0 -> NaN (impossible value)
n_bad_weight = (info['weight'] == 0).sum()
info.loc[info['weight'] == 0, 'weight'] = np.nan
print(f"Set {n_bad_weight} weight==0 values to NaN.")

# 3. Parse dates
for c in ['born_date', 'debut_date', 'last_date']:
    info[c] = pd.to_datetime(info[c], errors='coerce')
print("Converted born_date, debut_date, last_date to datetime dtype.")

# 4. Ensure id is a clean string key for merging later
info['id'] = info['id'].astype(str)

info_rows_after = len(info)
info_rows_before = len(info_df)
print(f"Rows before {info_rows_before}, and Rows after {info_rows_after}")

Removed 5 exact duplicate rows.
Set 2 weight==0 values to NaN.
Converted born_date, debut_date, last_date to datetime dtype.
Rows before 2848, and Rows after 2843


In [28]:
info.isnull().sum()

id                        0
player_name               0
player_full_name          0
first_name                0
last_name                 0
born_date                 0
debut_date                0
debut_age                 0
last_date                 0
last_age                  0
height                    0
weight                    2
profile_pic            2207
player_link               0
player_common_names    2768
player_teams             93
dtype: int64

### `seasonal_stats`

**Decisions:**
1. **Normalize `player_id`** — strip the `ID_` prefix from the 10 affected rows so every id is a plain numeric string matching `players_info.id`. Without this, those 10 rows would silently fail to merge.
2. **Drop exact duplicate rows** :identical player/year/team/stat combinations are double-entries, not separate seasons.
3. **Fix negative `games_played`**: a player cannot play a negative number of games. Taking the absolute value corrects what appears to be a sign-entry error (all other stats in these rows are positive and consistent with the corrected count).
4. **Standardize `team` names** :collapse case/whitespace variants (`' GEELONG CATS '`, `'geelong cats'`, `'Geelong Cats'`) into one canonical name matching the team naming convention used in `players_info`.
5. **Leave negative `total_fantasy_points` unchanged** :legitimate under AFL fantasy scoring (clangers, turnovers, and free kicks against subtract points), confirmed by the fact it only occurs in low-game-time rows.
6. **Leave missing advanced-stat values as `NaN`** : these reflect genuine historical non-tracking (e.g. hit-outs and Brownlow votes were not consistently recorded league-wide in earlier seasons). Imputing them (e.g. with 0 or mean) would fabricate performance data and bias any era-comparison analysis.


In [29]:
TEAM_MAP = {
    'adelaide crows': 'Adelaide Crows', 'brisbane bears': 'Brisbane Bears',
    'brisbane lions': 'Brisbane Lions', 'carlton blues': 'Carlton Blues',
    'collingwood magpies': 'Collingwood Magpies', 'essendon bombers': 'Essendon Bombers',
    'fitzroy lions': 'Fitzroy Lions', 'fremantle dockers': 'Fremantle Dockers',
    'geelong cats': 'Geelong Cats', 'gold coast suns': 'Gold Coast Suns',
    'greater western sydney giants': 'Greater Western Sydney Giants',
    'hawthorn hawks': 'Hawthorn Hawks', 'melbourne demons': 'Melbourne Demons',
    'north melbourne kangaroos': 'North Melbourne Kangaroos',
    'port adelaide power': 'Port Adelaide Power', 'richmond tigers': 'Richmond Tigers',
    'st kilda saints': 'St Kilda Saints', 'sydney swans': 'Sydney Swans',
    'west coast eagles': 'West Coast Eagles', 'western bulldogs': 'Western Bulldogs',
}

stats = stats_df.copy()

# 1. Normalize player_id (strip 'ID_' prefix)
n_prefixed = stats['player_id'].astype(str).str.contains('ID_').sum()
stats['player_id'] = stats['player_id'].astype(str).str.replace(r'\D', '', regex=True)
print(f" Normalized player_id: stripped 'ID_' prefix from {n_prefixed} values.")

# 2. Remove exact duplicate rows
dup_mask_s = stats.duplicated()
n_dup_s = dup_mask_s.sum()
stats = stats[~dup_mask_s].copy()
print(f" Removed {n_dup_s} exact duplicate rows.")

# 3. Fix negative games_played (sign error)
n_neg_games = (stats['games_played'] < 0).sum()
stats.loc[stats['games_played'] < 0, 'games_played'] = stats.loc[stats['games_played'] < 0, 'games_played'].abs()
print(f" Fixed {n_neg_games} rows with negative games_played (took absolute value).")

# 4. Standardize team names
before_unique = stats['team'].nunique()
stats['team'] = stats['team'].str.strip().str.lower().map(TEAM_MAP).fillna(stats['team'].str.strip())
after_unique = stats['team'].nunique()
print(f"Standardized team names: {before_unique} raw variants -> {after_unique} canonical teams.")

# 5. total_fantasy_points negatives -> left unchanged (legitimate)
n_neg_fp = (stats['total_fantasy_points'] < 0).sum()
print(f"total_fantasy_points negative in {n_neg_fp} rows - left unchanged (valid fantasy scoring outcome).")

stats_rows_after = len(stats)
stats_rows_before = len(stats_df)
print(f"Rows before {stats_rows_before}, and Rows after {stats_rows_after}")

 Normalized player_id: stripped 'ID_' prefix from 10 values.
 Removed 10 exact duplicate rows.
 Fixed 4 rows with negative games_played (took absolute value).
Standardized team names: 114 raw variants -> 20 canonical teams.
total_fantasy_points negative in 10 rows - left unchanged (valid fantasy scoring outcome).
Rows before 25491, and Rows after 25481


In [30]:
missing_summary = stats.isnull().sum()
missing_summary = missing_summary[missing_summary > 0]
print(f"[stats] {len(missing_summary)} columns retain missing values:- left as NaN")
missing_summary.sort_values(ascending=False)


[stats] 45 columns retain missing values:- left as NaN


goal_assists                   7092
avg_goal_assists               7092
brownlow_votes                 6997
total_percentage_played        6245
avg_percentage_played          6245
avg_hit_outs                   5706
hit_outs                       5706
bounces                        5654
avg_bounces                    5654
avg_marks_inside_50            5165
marks_inside_50                5165
contested_marks                4925
avg_contested_marks            4925
one_percenters                 3643
avg_one_percenters             3643
avg_rebound_50s                3533
rebound_50s                    3533
clearances                     3435
avg_clearances                 3435
contested_possessions          3336
avg_contested_possessions      3336
avg_uncontested_possessions    3322
uncontested_possessions        3322
avg_inside_50s                 2971
inside_50s                     2971
clangers                       2938
avg_clangers                   2938
avg_behinds                 

In [31]:
count_cols_with_na = [c for c in stats.columns 
                       if stats[c].isnull().sum() > 0 and not c.startswith('avg_')]

stats[count_cols_with_na] = stats[count_cols_with_na].fillna(0)

# Recompute the matching avg_ column from the (now-filled) count and games_played,rather than filling avg_ columns independently
for col in count_cols_with_na:
    avg_col = f'avg_{col}'
    if avg_col in stats.columns:
        stats[avg_col] = (stats[col] / stats['games_played']).round(1)

remaining_avg_na = [c for c in stats.columns if c.startswith('avg_') and stats[c].isnull().sum() > 0]
stats[remaining_avg_na] = stats[remaining_avg_na].fillna(0)

print(f"Filled {len(count_cols_with_na)} count columns and recomputed their averages.")
print(f"Filled {len(remaining_avg_na)} remaining avg-only columns with 0.")
print("Remaining NaNs in stats:", stats.isnull().sum().sum())

Filled 23 count columns and recomputed their averages.
Filled 1 remaining avg-only columns with 0.
Remaining NaNs in stats: 0


Did this because those are old records, they dont need stuff so filled it with 0

In [32]:
stats.isnull().sum()

player_id                      0
year                           0
team                           0
is_finals                      0
games_played                   0
kicks                          0
marks                          0
handballs                      0
disposals                      0
goals                          0
behinds                        0
hit_outs                       0
tackles                        0
rebound_50s                    0
inside_50s                     0
clearances                     0
clangers                       0
free_kicks_for                 0
free_kicks_against             0
brownlow_votes                 0
contested_possessions          0
uncontested_possessions        0
contested_marks                0
marks_inside_50                0
one_percenters                 0
bounces                        0
goal_assists                   0
total_score                    0
total_fantasy_points           0
total_percentage_played        0
avg_kicks 

## 3. Merge & Unmatched-ID Analysis

Players are merged on `seasonal_stats.player_id` = `players_info.id`. We use an
**inner join** for the analysis-ready `merged_players.csv` (every row has both
season performance and player biographical data), and separately report every
`player_id` that failed to match so no data loss is hidden.

In [33]:
info_ids = set(info['id'])
stats_ids = set(stats['player_id'])

unmatched_in_stats = stats_ids - info_ids   # season-stat ids with no player record
unmatched_in_info = info_ids - stats_ids    # players with no season stats at all

print(f"Unique player ids - stats: {len(stats_ids)}, info: {len(info_ids)}")
print(f"player_id in seasonal_stats with NO match in players_info: {len(unmatched_in_stats)}")
print(f"id in players_info with NO season stats recorded: {len(unmatched_in_info)}")


Unique player ids - stats: 3109, info: 2843
player_id in seasonal_stats with NO match in players_info: 266
id in players_info with NO season stats recorded: 0


In [34]:
merged = stats.merge(info, left_on='player_id', right_on='id', how='inner', suffixes=('', '_info'))
merged = merged.drop(columns=['id'])  # redundant with player_id

n_merged = len(merged)
print(f"row count: {n_merged} "
    f"({stats_rows_after - n_merged} stats rows excluded :unmatched player_id).")

merged.head()


row count: 25081 (400 stats rows excluded :unmatched player_id).


,player_id,year,team,is_finals,games_played,kicks,marks,handballs,disposals,goals,...,debut_date,debut_age,last_date,last_age,height,weight,profile_pic,player_link,player_common_names,player_teams
0,43261,2018,Geelong Cats,False,3,17.0,9.0,21.0,38.0,3.0,...,2018-08-02,27,2020-09-05,29,200,100.0,NaN,https://afltables.com/afl/stats/players/R/Ryan...,NaN,"{Geelong Cats,St Kilda Saints}"
1,43261,2018,Geelong Cats,True,1,3.0,2.0,2.0,5.0,0.0,...,2018-08-02,27,2020-09-05,29,200,100.0,NaN,https://afltables.com/afl/stats/players/R/Ryan...,NaN,"{Geelong Cats,St Kilda Saints}"
2,43261,2019,Geelong Cats,False,1,5.0,3.0,6.0,11.0,1.0,...,2018-08-02,27,2020-09-05,29,200,100.0,NaN,https://afltables.com/afl/stats/players/R/Ryan...,NaN,"{Geelong Cats,St Kilda Saints}"
3,43261,2020,St Kilda Saints,False,1,2.0,2.0,1.0,3.0,1.0,...,2018-08-02,27,2020-09-05,29,200,100.0,NaN,https://afltables.com/afl/stats/players/R/Ryan...,NaN,"{Geelong Cats,St Kilda Saints}"
4,43262,2002,Geelong Cats,False,12,37.0,13.0,63.0,100.0,10.0,...,2002-03-30,17,2020-10-24,36,182,87.0,NaN,https://afltables.com/afl/stats/players/G/Gary...,NaN,"{Geelong Cats,Gold Coast Suns}"


In [35]:
unmatched_ids_df = pd.DataFrame(sorted(unmatched_in_stats), columns=['unmatched_player_id'])
print(f"Sample of unmatched player_id values ({len(unmatched_ids_df)} total):")
unmatched_ids_df.head(20)


Sample of unmatched player_id values (266 total):


,unmatched_player_id
0,43280
1,43292
2,43299
3,43319
4,43325
5,43329
6,43332
7,43348
8,43359
9,43361


## 4. Validation Report

Row counts, missing-value handling, duplicates removed, and unmatched ids

In [36]:
validation_report = f"""
AFL DATA VALIDATION REPORT
{'='*60}

ROW COUNTS
  players_info     : {info_rows_before} -> {info_rows_after}  ({info_rows_before - info_rows_after} removed)
  seasonal_stats    : {stats_rows_before} -> {stats_rows_after}  ({stats_rows_before - stats_rows_after} removed)
  merged_players    : {n_merged} rows (inner join of cleaned datasets)

DUPLICATE RECORDS REMOVED
  players_info exact duplicate rows removed   : {n_dup}
  seasonal_stats exact duplicate rows removed : {n_dup_s}

MISSING VALUES HANDLED
  players_info.weight == 0 -> set to NaN      : {n_bad_weight} values
  players_info optional fields left as NaN    : profile_pic, player_common_names, player_teams
  seasonal_stats advanced-stat NaNs left as-is: {len(missing_summary)} columns (since its era based, wasnt tracked)

OTHER CORRECTIONS
  seasonal_stats.player_id 'ID_' prefix normalized : {n_prefixed} values
  seasonal_stats.games_played negative -> abs()     : {n_neg_games} rows
  seasonal_stats.team name variants standardized    : {before_unique} -> {after_unique} names

UNMATCHED player_id VALUES AFTER MERGING
  seasonal_stats ids with no players_info match : {len(unmatched_in_stats)}
  players_info ids with no seasonal_stats match : {len(unmatched_in_info)}
  -> {stats_rows_after - n_merged} seasonal_stats rows excluded from merged_players.csv as a result
"""
print(validation_report)



AFL DATA VALIDATION REPORT

ROW COUNTS
  players_info     : 2848 -> 2843  (5 removed)
  seasonal_stats    : 25491 -> 25481  (10 removed)
  merged_players    : 25081 rows (inner join of cleaned datasets)

DUPLICATE RECORDS REMOVED
  players_info exact duplicate rows removed   : 5
  seasonal_stats exact duplicate rows removed : 10

MISSING VALUES HANDLED
  players_info.weight == 0 -> set to NaN      : 2 values
  players_info optional fields left as NaN    : profile_pic, player_common_names, player_teams
  seasonal_stats advanced-stat NaNs left as-is: 45 columns (since its era based, wasnt tracked)

OTHER CORRECTIONS
  seasonal_stats.player_id 'ID_' prefix normalized : 10 values
  seasonal_stats.games_played negative -> abs()     : 4 rows
  seasonal_stats.team name variants standardized    : 114 -> 20 names

UNMATCHED player_id VALUES AFTER MERGING
  seasonal_stats ids with no players_info match : 266
  players_info ids with no seasonal_stats match : 0
  -> 400 seasonal_stats rows exclud

# Saving cleaned data

In [37]:
info.to_csv('players_info_cleaned.csv', index=False)
stats.to_csv('seasonal_stats_cleaned.csv', index=False)
merged.to_csv('merged_players.csv', index=False)
unmatched_ids_df.to_csv('unmatched_player_ids.csv', index=False)



# Overall Findings
### **players_info had** :
- 5 exact duplicate rows
- and also2 impossible weight=0 values
### **seasonal_stats had** :
- inconsistent player_id formats
- 10 duplicate rows
- 4 rows with negative games_played
- 114 raw team-name string variants collapsed to 20 teams.
- Missing values in advanced stats (hit-outs, Brownlow votes) were filled with 0 since they reflect that those eras werent tracked.

**After cleaning, 266 player_id values in seasonal_stats had no match in players_info, those are documented in unmatched_player_ids.csv and excluded from the inner-joined merged_players.csv which has 25081 rows**

In [39]:
info.shape

(2843, 16)

In [40]:
info_df.shape

(2848, 16)

In [41]:
stats_df.shape

(25491, 54)

In [42]:
stats.shape

(25481, 54)